### PR TITLE
Extending BasicTracerProvider

### PR DESCRIPTION
**Pull request by kolanos**

### CODE REVIEW

File: azureLogIngestion/opentelemetry/index.ts [hunk 1]
```diff
@@ -96,7 +98,7 @@ export default class OpenTelemetryAdapter {
                     : "https://otlp.nr-data.net:4317/v1/traces",
         })
 
-        this.traceProvider = new BasicTracerProvider({
+        this.traceProvider = new NRTracerProvider({
```
Comment (mrickard) [inline azureLogIngestion/opentelemetry/index.ts:101]: The problem here is going to be that we're setting the resource before the first span, when we don't have an actual resource yet. 

What happens when you invoke .updateResource from within addSpan?

File: azureLogIngestion/opentelemetry/provider.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import { Resource } from "@opentelemetry/resources"
+import { BasicTracerProvider } from "@opentelemetry/tracing"
+
+export class NRTracerProvider extends BasicTracerProvider {
+    resource: Resource
```
Comment (mrickard) [inline azureLogIngestion/opentelemetry/provider.ts:5]: For this class definition, the typescript hinting complains that `TS2610: 'resource' is defined as an accessor in class 'BasicTracerProvider', but is overridden here in 'NRTracerProvider' as an instance property.`

But if it tests and builds...we might be able to ignore that. (That's what led me down the path of trying to work with the private `_resource` property.

Comment (mrickard) [reply azureLogIngestion/opentelemetry/provider.ts:5]: Also, when I run the test, I'm getting that TS2610 error.

**Comment by mrickard** [comment]
Unfortunately, when I run build on this branch, the build fails with two errors: 

`TS2610: 'resource' is defined as an accessor in class 'BasicTracerProvider', but is overridden here in 'NRTracerProvider' as an instance property.`

`TS1183: An implementation cannot be declared in ambient contexts.`

Unless there's a difference in our build environments, we're going to have to find another way around this.

**Review by mrickard — CHANGES_REQUESTED** [review]
Unfortunately, when I run build on this branch, the build fails with two errors: 

`TS2610: 'resource' is defined as an accessor in class 'BasicTracerProvider', but is overridden here in 'NRTracerProvider' as an instance property.`

`TS1183: An implementation cannot be declared in ambient contexts.`

Unless there's a difference in our build environments, we're going to have to find another way around this.

File: azureLogIngestion/opentelemetry/index.ts (hunk 1 above)
Comment (kolanos) [reply azureLogIngestion/opentelemetry/index.ts:101]: You will need to look at the implementation of `BasicTracerProvider`, `Tracer` and `Span` to understand how this works. In short, when you start a Tracer, a bunch of meta data, including the resource, get passed to the Tracer from the TracerProvider, which in turn gets passed to the Span when a span is started. We get a Tracer instance in `addSpan`, then start/end a span. Whatever resource is set in the tracer provider is used and added to the batch.

File: azureLogIngestion/opentelemetry/provider.ts (hunk 2 above)
Comment (kolanos) [reply azureLogIngestion/opentelemetry/provider.ts:5]: I'm running `npm run build` without error. Is this some IDE you're running? That may be enabling additional linting.

File: azureLogIngestion/opentelemetry/index.ts (hunk 1 above)
Comment (mrickard) [reply azureLogIngestion/opentelemetry/index.ts:101]: Yeah, I've been digging in there. Both the `Tracer` and `BasicTracerProvider` have private readonly `_resource` properties with a `resource` accessor.  I haven't yet figured out a way of overriding those that doesn't throw a build error.

File: azureLogIngestion/opentelemetry/provider.ts (hunk 2 above)
Comment (mrickard) [reply azureLogIngestion/opentelemetry/provider.ts:5]: I'm running my builds also with `npm run build` in the terminal from within IntelliJ. I do get the same build errors from my default terminal, though. My test branch fails on build in github, though curiously this PR doesn't seem to have run that same build step.

**Review by mrickard — APPROVED** [review]

**Pull request merged**